### ISSUE TITLE
Page and images not fully loaded before actions taken

### ISSUE THREAD

**Issue by chunsli** [webdriverio/webdriverio#1880]
## The problem

The page is occasionally not fully loaded before screenshots are taken. Most of the times it works.

## Environment

* WebdriverIO version: 4.6.2
* Node.js version: 6.9.2
* Standalone version of WebdriverIO
* Browser version: Google Chrome 55.0.2883.87
## Details

Expected output (most of the times this is the actual output):
![9uf_4 template1](https://cloud.githubusercontent.com/assets/11830348/23011354/8a9127de-f45b-11e6-8b4f-a2fdf1cc96a1.png)

Misbehaving output:
![9uf_4 template1](https://cloud.githubusercontent.com/assets/11830348/23011333/634a32f6-f45b-11e6-8c09-1c462fc8b790.png)

## Link to Selenium/WebdriverIO logs
07:20:38.097 INFO - Done: [new session: Capabilities [{rotatable=true, locationContextEnabled=true, loggingPrefs=org.openqa.selenium.logging.LoggingPreferences@7f775313, browserName=chrome, javascriptEnabled=true, handlesAlerts=true, requestOrigins={name=webdriverio, version=4.6.2, url=http://webdriver.io}}]]
07:20:38.271 INFO - Executing: [get: file:///home/usr/test/uploads/html/11092178/1487229635013-9UF_4(template1).html])
07:20:39.555 INFO - Executing: [execute script: return (function () {
        var imgs = document.getElementsByTagName("img");
        if (imgs.length > 0) {
          Array.prototype.every.call(imgs, function (img) {
            return img.complete;
          });
        } else {
          return true;
        }
      }).apply(null, arguments), []])
07:20:39.812 INFO - Executing: [take screenshot])

## Code To Reproduce Issue [ Good To Have ]
(Not sure if related, but I am running (~10) multiple browsers in parallel for each batch)
```
  const options = {
    desiredCapabilities: {
      browserName: 'chrome'
    },
    port: 4444
  };
  const browser = webdriverio.remote(options);
  browser.init()
    .windowHandleSize({width: 1280, height: 1280})
    .url('file://' + path.resolve(__dirname, '../../../uploads/html/', filePath))
    .waitUntil(function() {
      return this.execute(function(){
        let imgs = document.getElementsByTagName("img");
        if (imgs.length > 0) {
          Array.prototype.every.call(imgs, img => {
            return img.complete;
          })
        } else {
          return true;
        }
      });
    },100000,`CRITICAL ERROR: process image time out at ${filename}`, 1000)
    .saveScreenshot(filename)
    .end()
```
Please remember that, with sample code; it's easier to reproduce bug and much faster to fix it.

**Comment by christian-bromann** [comment]
I am not sure what the issue here is. You already use a script to make sure the images are loading properly before taking the screenshot. I think this is the right approach.

**Comment by chunsli** [comment]
I think what happened was that images were not available when waitUntil was executed, so it automatically returned true, I added document.readyState === 'complete' to check that the document is loaded.

However, what still confuses me is that why document.readyState === 'complete' alone won't be able to check for the readiness of the images. It still produces incomplete images if used by itself.
```
  browser.init()
    .windowHandleSize({width: 1280, height: 1280})
    .url('file://' + path.resolve(__dirname, '../../../uploads/html/', filePath))
    .waitUntil(function() {
      return this.execute(function(){
        let imgs = document.getElementsByTagName("img");
        if (imgs.length > 0) {
          return Array.prototype.every.call(imgs, img => {
            return img.complete;
          }) && document.readyState === 'complete'
        } else {
          return true;
        }
      });
    },100000,`CRITICAL ERROR: process image time out at ${filename}`, 3000)
    .saveScreenshot(filename)
    .end()
```

**Comment by mattc41190** [comment]
@chunsli not sure if this will help, but maybe I am understanding what is confusing you.

https://4loc.wordpress.com/2009/04/28/documentready-vs-windowload/

**Comment by chunsli** [comment]
@mattc41190 

Thanks for helping. However, I doubt this is the issue. Let's do a simple test with this simple HTML code, which basically shows you when `$(document).ready`, `document.readyState === 'complete'` and `img.complete` fire:

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
  <script>
    let id;
    function imageComplete(){
      let imgs = document.getElementsByTagName("img");
      if (imgs.length > 0) {
        return Array.prototype.every.call(imgs, img => {
          return img.complete;
        })
      } else {
        return true;
      }
    }
    function readyState(){
      return document.readyState === 'complete';
    }
    $(document).ready(function(){
      console.log('document is ready')
    });
    id = setInterval(checker, 500);
    function checker () {
      console.log('image.complete', imageComplete());
      console.log('readyState', readyState());
      if(imageComplete() && readyState()) {
        clearInterval(id);
      }
    }
  </script>
</head>
<body>
<img src="http://lorempixel.com/1024/1024/" />
</body>
</html>
```

As expected, jQuery's $(document).ready fires before images are loaded. However, document.readyState turns to complete at the same time img.complete turns true. (I throttled the network to make results more obvious) 

![image](https://cloud.githubusercontent.com/assets/11830348/23111881/8dc02348-f766-11e6-92b0-4cb2a6293eb8.png)

I did spot some issues with my code, however, the result is still suboptimal (`this.execute` returns a Promise, I need to create an async function):

 ```
 browser.init()
    .windowHandleSize({width: 1280, height: 1280})
    .url('file://' + path.resolve(__dirname, '../../../uploads/html/', filePath))
    .waitUntil(async function() {
      const result = await this.execute(function(){
        let imgs = document.getElementsByTagName("img");
        if (imgs.length > 0) {
          return Array.prototype.every.call(imgs, img => {
            return img.complete;
          }) && document.readyState === 'complete'
        } else {
          return true;
        }
      });
      console.log('result', result);
      return result;
    },100000,`CRITICAL ERROR: process image time out at ${filename}`, 3000)
    .saveScreenshot(filename)
    .end()
```

**Comment by HouCoder** [comment]
This is my approach:


```js
it('scroll right and left', () => {
    // Wait for document to be fully loaded.
    browser.waitUntil(
        () => {
            const state = browser.execute(function () {
                return document.readyState;
            });
            return state.value === 'complete';
        },
        100000,
        'Oops, time out!',
        500
    );

    // Your custom actions...
```

**Comment by chunsli** [comment]
Thanks for the response, I have given up on webdriverIO, and opting to use [puppeteer](https://github.com/GoogleChrome/puppeteer).

There are also other choices like [nightmare](https://github.com/segmentio/nightmare)

The advantage with puppeteer is that it doesn't require installation of native binaries, a simple `npm install` will do, and all functions work like a charm. Highly recommend switching over.

**Comment by christian-bromann** [comment]
> Highly recommend switching over.

@chunsli this might be true for your use case but definitely not in general.
